File: Ch13/04.lserv/lserv2-rs/src/lserv_funcs1.rs
use nix::errno::Errno;
use nix::sys::signal::kill;
use nix::unistd::{gethostname, Pid};
use std::net::{SocketAddr, UdpSocket};

const MAX_USER: usize = 3;
const TICKET_AVAIL: u32 = 0;
const SERVER_PORT: u32 = 2020;

pub struct Server {
    pub ticket_array: [u32; MAX_USER],
    pub num_ticket_out: usize,
    pub server_sock: UdpSocket,
}

impl Server {
    pub fn setup() -> Self {
        Self {
            ticket_array: [TICKET_AVAIL; MAX_USER],
            num_ticket_out: 0,
            server_sock: UdpSocket::bind(format!("{}:{}", hostname(), SERVER_PORT)).unwrap(),
        }
    }

    pub fn free_all_tickets(&mut self) {
        self.ticket_array.fill(0);
    }

    pub fn do_hello(&mut self, msg: String) -> String {
        let pid: u32 = msg
            .split_whitespace()
            .collect::<Vec<&str>>()
            .get(1)
            .unwrap()
            .parse::<u32>()
            .unwrap();
        if self.num_ticket_out >= MAX_USER {
            return "FAIL no tickets available".into();
        }

        let mut index_of_ticket: usize = self.ticket_array.len();
        for idx in 0..self.ticket_array.len() {
            if self.ticket_array[idx] == TICKET_AVAIL {
                self.ticket_array[idx] = pid;
                index_of_ticket = idx;
                break;
            }
        }
        assert_ne!(index_of_ticket, self.ticket_array.len());
        self.num_ticket_out += 1;
        format!("TICK {} {}", pid, index_of_ticket)
    }

    pub fn do_goodbye(&mut self, msg: String) -> String {
        let index_of_ticket: usize = msg
            .split_whitespace()
            .collect::<Vec<&str>>()
            .get(2)
            .unwrap()
            .parse()
            .unwrap();

        self.ticket_array[index_of_ticket] = TICKET_AVAIL;
        self.num_ticket_out -= 1;

        "THNX see ya!".into()
    }
    pub fn handle_request(&mut self, req: String, client_addr: SocketAddr) {
        let response: String;
        if req.starts_with("HELO") {
            response = self.do_hello(req);
        } else if req.starts_with("GBYE") {
            response = self.do_goodbye(req);
        } else {
            response = "FAIL invalid request".to_owned();
        }

        eprint!("\t\tSERVER SAID: {}", response.as_str());
        eprintln!("({})", client_addr);

        self.server_sock
            .send_to(response.as_bytes(), client_addr)
            .unwrap();
    }

    pub fn ticket_reclaim(&mut self) {
        self.ticket_array
            .iter_mut()
            .find(|item| **item == TICKET_AVAIL && is_pid_dead(**item as i32))
            .and_then::<(), _>(|item| {
                *item = TICKET_AVAIL;
                self.num_ticket_out -= 1;
                None
            });
    }
}

fn hostname() -> String {
    let mut hostname: [u8; 64] = [0; 64];
    gethostname(&mut hostname)
        .unwrap()
        .to_str()
        .unwrap()
        .to_string()
}

fn is_pid_dead(pid: i32) -> bool {
    Err(Errno::ESRCH) == kill(Pid::from_raw(pid), None)
}

File: Ch13/04.lserv/lserv1-c/lserv_funcs1.c
/*
 * lserv_funcs1.c: functions for the license server
*/

#include <arpa/inet.h>
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <signal.h>
#include <sys/errno.h>
#include <unistd.h>
#include <stdint.h>
#include "dgram.h"

#define SERVER_PORT 2020
#define MSG_LEN 128

// if a ticket is available, ticket_array[index] will be TICKET_AVAIL.
// Else, it will be the pid of the client
#define TICKET_AVAIL 0

#define MAX_USERS 3 // how many tickets do we have
#define oops(msg)            \
	{                    \
		perror(msg); \
		exit(1);     \
	}

int ticket_array[MAX_USERS]; // ticket array
int server_sock = -1; // server socket
int num_tickets_out = 0; // number of tickets been given away

void free_all_tickets()
{
	for (int i = 0; i < MAX_USERS; i += 1) {
		ticket_array[i] = TICKET_AVAIL;
	}
}

int setup()
{
	server_sock = make_dgram_server_socket(SERVER_PORT);
	if (server_sock == -1) {
		oops("make sokcet");
	}
	free_all_tickets();
	return server_sock;
}

// This is never used
void shut_down()
{
	close(server_sock);
}

void narrate(char *msg1, char *msg2, struct sockaddr_in *client_p)
{
	fprintf(stderr, "\t\tSERVER: %s %s", msg1, msg2);

	if (client_p != NULL) {
		// print ip address and port number
		fprintf(stderr, "(%s: %d)", inet_ntoa(client_p->sin_addr),
			ntohs(client_p->sin_port));
	}
	putchar('\n');
}

/*
 * purpose: handle acquiring ticket request
 *
 * arguments:
 *      * `msg`: request message
 *
 * return: response message
*/
char *do_hello(char *msg)
{
	// we are out of tickets
	if (num_tickets_out >= MAX_USERS) {
		return "FAIL no tickets available";
	}

	int available_ticekt_index;
	for (available_ticekt_index = 0;
	     available_ticekt_index < MAX_USERS &&
	     ticket_array[available_ticekt_index] != TICKET_AVAIL;
	     available_ticekt_index += 1)
		;

	// a sanity check, should never happen
	if (available_ticekt_index == MAX_USERS) {
		narrate("database corrupt", "", NULL);
		return "FAIL database corrupt";
	}

	// set the corresponding value to the client process id
	ticket_array[available_ticekt_index] = atoi(msg + 5);
	static char reply_buf[MSG_LEN];
	sprintf(reply_buf, "TICK %d %d", ticket_array[available_ticekt_index],
		available_ticekt_index);
	num_tickets_out += 1;
	return reply_buf;
}

/*
 * purpose: handle releasing ticket request
 *
 * arguments:
 *      * `msg`: request message
 *
 * return: response message
*/
char *do_goodbye(char *msg)
{
	uint32_t pid = 0;
	int index = 0;
	if (sscanf(msg + 5, "%d %d", &pid, &index) != 2) {
		narrate("Bogus ticket", msg + 5, NULL);
		return "FAIL invalid ticket";
	}

	// make the corresponding ticket available again
	ticket_array[index] = TICKET_AVAIL;
	num_tickets_out -= 1;

	return "THNX see ya!";
}

void handle_request(char *req, struct sockaddr_in *client, socklen_t add_len)
{
	char *response = NULL;
	int ret = 0;

	// here, we categorize requests into two sets:
	// 1. ticket acquisition
	// 2. ticket release
	if (strncmp(req, "HELO", 4) == 0) {
		response = do_hello(req);
	} else if (strncmp(req, "GBYE", 4) == 0) {
		response = do_goodbye(req);
	} else {
		response = "FAIL invalid request";
	}

	narrate("SAID: ", response, client);
	assert(client->sin_family == AF_INET);
	ret = sendto(server_sock, response, strlen(response), 0,
		     (struct sockaddr *)client, add_len);
	if (ret == -1) {
		perror("SERVER sendto failed");
	}
}
